src/legacy/core_plugins/vega/public/plugin.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// @ts-ignore
+import { createVegaFn } from './vega_fn';
+// @ts-ignore
+import { createVegaTypeDefinition } from './vega_type';
+
+import { VegaSetupPlugin } from './setup';
+
+export class VegaPlugin {
+  public async setup({ data, visualizations, legacy }: VegaSetupPlugin) {
```

Comment:
I like this approach where all `legacy` stuff is injected into the shim instead of imported from `ui/...`.

---

@elastic/kibana-app-arch Let's discuss it on Tuesday sync.

I would rename it to something like `__LEGACY` and it would need to be a "plugin", not service. But I somehow like this much more than arbitrarily importing from `ui/...` from various files within a shim. This way you explicitly see what legacy stuff you plugin still depends on.

Maybe we could have a convention where shims are in `<legacy_plugin>/public/shim` folder and add a linter rule that does not allow to import anything from `ui/...` in any of the `<legacy_plugin>/public/shim/public/**` files.

Reply:
Yeah that's certainly an option too. To prevent confusion, we should move this discussion over to #39554 and can report back here